contrib/bloop/src/mill.contrib.bloop/BloopModule.scala

```diff
@@ -0,0 +1,197 @@
+package mill.contrib.bloop
+
+import ammonite.ops._
+import bloop.config.Config.{File => BloopFile}
+import bloop.config.ConfigEncoderDecoders._
+import io.circe.Encoder
+import mill._
+import mill.define._
+import mill.eval.Evaluator
+import mill.modules.Jvm
+import mill.scalalib._
+import ujson.Visitor
+import upickle.default
+
+// TODO diamond problem is going to make this tricky. Using a self-type appears
+// tobe problematic with mill. Moreover, the global "install"
+// method that inspects the build relies on whether the modules are instances of
+// JavaModule or ScalaModule, so not quite sure what approach to take to avoid
+// copy pasting between a "BloopJavaModule" and a "BloopScalaModule" (and a potential
+// BloopScalaJsModule ?)
+//
+// TODO : bloop clean doesn't seem to work for me. Also, investigate if it's
+// possible to hook with "mill clean"
+//
+trait BloopModule extends ScalaModule { self =>
+
+  trait Tests extends super.Tests with BloopModule {
+    // TODO how can we retrieve test results from bloop ?
```

Comment:
I've been giving this a little bit of thought, and the best way to retrieve test results (but also compilation results) from bloop would be by using BSP. Mill would be a BSP client and would connect to Bloop via BSP. https://github.com/scalacenter/bsp/blob/master/docs/bsp.md#test-request